### PR TITLE
[BACKLOG-9117] Automatic axis label's non-overlapping angle

### PR DESCRIPTION
**Pull request by nantunes**
@pentaho/millenniumfalcon, please review.

@dcleao, I had conflits and needed to rebase, so you'll need to also rebase your branch.

@jheck123, you just need to comment on commit 378c778.

### CODE REVIEW

**Comment by nantunes** [comment]
@pentaho/millenniumfalcon => @dcleao @graimundo @davidmsantos90 @carlosrusso @marcovala

File: doc/model/axes/axis-cartesian-discrete.xml [hunk 1]
```diff
@@ -32,6 +32,25 @@
             For additional information, please see <c:link to="pvc.options.axes.CartesianAxis" />.
         </c:documentation>
 
+        <c:property name="labelRotationDirection" 
+                    type="pvc.options.varia.AxisLabelRotationDirection"
+                    default="'clockwise'"
+                    category="Non-Hierarchical > Layout">
+            <c:documentation>
+                Controls in which direction labels rotate.
```
Comment (dcleao) [inline doc/model/axes/axis-cartesian-discrete.xml]: Please include the suffix: "Controls in which direction labels rotate (applies to discrete axes)."

File: doc/model/axes/axis-cartesian-discrete.xml [hunk 2]
```diff
@@ -32,6 +32,25 @@
...
+            </c:documentation>
+        </c:property>
+
+        <c:property name="labelDesiredAngles" 
+                    type="number string list(number string)"
+                    default="[]"
+                    category="Non-Hierarchical > Layout">
+            <c:documentation>
+                List of angles to try when overlappedLabelsMode is 'rotate' or 'rotatethenhide'.
```
Comment (dcleao) [inline doc/model/axes/axis-cartesian-discrete.xml]: Please include the suffix "... 'rotatethenhide' (applies to discrete axes)." and a blank line between the two statements.

File: doc/model/axes/axis-cartesian-discrete.xml [hunk 3]
```diff
@@ -32,6 +32,25 @@
             For additional information, please see <c:link to="pvc.options.axes.CartesianAxis" />.
         </c:documentation>
 
+        <c:property name="labelRotationDirection"
```
Comment (dcleao) [inline doc/model/axes/axis-cartesian-discrete.xml]: These two options also need to be included in a separate complex type, `AnyNonHierarchicalCartesianAxis`. To be able to reuse these in both types, these should be moved into a facet type, for example: `FlattenedDiscreteCartesianAxisFacet`. Then this facet is referenced by both of the concrete complex types (in their `facets` attribute): `FlattenedDiscreteCartesianAxis` and `AnyNonHierarchicalCartesianAxis`.

**Comment by nantunes** [comment]
@dcleao please recheck the doc facet issue. ty

**Comment by dcleao** [comment]
@nantunes regarding moving the docs to `CartesianAxisTicksFacet`, good catch. Unfortunately, that facet is also used by continuous axis complex types: `AnyContinuousCartesianAxis` and `NumericCartesianAxis`. So your move was good for two reasons:
1. We've just found a name for the new facet that is closer to existing names: `CartesianAxisTicksFacet`
2. We detected a bug in the docs, where the `overlappedLabelsMode` options is incorrectly being included in continuous axes... although it only applies to non-hierarchical discrete axes... the same case of the new options we're now adding :-)

File: package-res/ccc/core/cartesian/axis/cart-axis.js [hunk 4]
```diff
@@ -590,6 +590,22 @@ function cartAxis_castTitleSize(value) {
     return pvc_Size.to(value, {singleProp: pvc.BasePanel.orthogonalLength[position]});
 }
 
+function cartAxis_labelDesiredAngles(value) {
+    if(!value) {
+        return [];
+    }
+
+    if(value.constructor !== Array) {
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/cart-axis.js]: Please use: `!def.array.is(value)`

File: package-res/ccc/core/cartesian/axis/cart-axis.js [hunk 5]
```diff
@@ -590,6 +590,22 @@ function cartAxis_castTitleSize(value) {
...
+        value = [value];
+    }
+
+    for (var i = 0, ic = value.length; i != value.length; ++i) {
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/cart-axis.js]: i !== ic ?

File: package-res/ccc/core/cartesian/axis/cart-axis.js [hunk 6]
```diff
@@ -590,6 +590,22 @@ function cartAxis_castTitleSize(value) {
...
+        value[i] = def.number.to(value[i]);
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/cart-axis.js]: when value[i] is not a number, `def.number.to` is returning undefined. We should probably filter these out and push into an output array instead.

File: package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js [hunk 7]
```diff
@@ -1115,9 +1056,25 @@ def
...
             .font(font)
-            .textStyle("#666666")
-            .textAlign(layoutInfo.textAlign)
-            .textBaseline(layoutInfo.textBaseline);
+            .textStyle("#666666");
+
+        if(layoutInfo.lockedTextAngle !== undefined) {
+            this.pvLabel.lock('textAngle', layoutInfo.lockedTextAngle);
+        } else {
+            this.pvLabel.textAngle(layoutInfo.textAngle)
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js]: ";" missing and the same in the following ifs

File: package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js [hunk 8]
```diff
@@ -1553,3 +1510,281 @@ def
...
+ *                                       ticks.length:       The axis total number of ticks.
+ *
+ *                                      This function always adds these properties:
+ *                                       tickVisibilityStep: The visibility step of the labels.
+ *                                       lockedTextAngle:    The text angle to apply to the labels.
+ *
+ *                                      Additionally, if overlappedLabelsMode is 'rotate' or 'rotatethenhide', these
+ *                                      properties are also defined:
+ *                                       lockedTextAlign:    The horizontal label anchor.
+ *                                       lockedTextBaseline: The vertical label anchor.
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js]: Following the naming used in similar cases, like `beginLocked`, `leftLocked`, `minLocked`, etc.. , these would be named `textAlignLocked` and `textBaselineLocked`.

File: package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js [hunk 9]
```diff
@@ -1553,3 +1510,281 @@ def
...
+            // if none, chooses to the last desired angle
+            for (var i = 0; i !== labelDesiredAngles.length; ++i) {
+                a = pvc.normAngle(labelDesiredAngles[i] * labelRotationDirection);
+
+                // angular distance to the closest axis
+                var abs_angle = a > FLAT_ANGLE ? Math.abs(a - FULL_ANGLE) : a;
+                var angle_to_axis = abs_angle < RIGHT_ANGLE ? abs_angle : abs_angle - FLAT_ANGLE;
+                if(angle_to_axis >= min_angle) {
+                    // no need for hiding if between min and max angle, it will never overlap
+                    canHide = canHide && (isNaN(max_angle) || angle_to_axis > max_angle);
```
Comment (dcleao) [inline package-res/ccc/core/cartesian/axis/abstract-cart-axis-panel.js]: By what I could understand, the `NaN` in min_angle and max_angle can only occur due to the arguments of `Math.asin` or `Math.acos` being out of range (-1, 1). What situations do these correspond to? Do they actually happen?
It feels strange to me that we don't guard against `NaN` occurrence closer to the "offending" calculations.
So, if in this line we guard against `max_angle` being `NaN`, why don't we do the same for the previous line, when using `min_angle`? Or are we relying on that comparing to `NaN` never returns `true`? In that case, why bother determining the value of `angle_to_axis`?

Mostly, I'd like that the comments would give some clue of what "use case" NaN corresponds to, if any. That would provide meaning for the following `isNaN` tests.

**Comment by dcleao** [comment]
👍

**Comment by nantunes** [comment]
@dcleao PR updated with your feedback. Will think about the label anchoring algorithm later.
Also, should we do something related to the facet?

**Comment by dcleao** [comment]
@nantunes regarding the facet, I just read what I wrote, and I spotted a mistake that might have generated some confusion. This is what I wanted to say:

"We've just found a name for the new facet that is closer to existing names: **Discrete**CartesianAxisTicksFacet"

So, yes, I think we should create a new facet `DiscreteCartesianAxisTicksFacet`, where the two new properties will be placed, and the `overlappedLabelsMode` property will be moved to.

We'd fix the category attributes to "Discrete > Layout" and the suffix in the description paragraphs to something like "(applies to discrete, non-hierarchical axes)" or "(applies to normal discrete axes)" ?

Then we'd add this new facet to the discrete axis complex types only: `DiscreteCartesianAxis` and `AnyDiscreteCartesianAxis`, beside the existing reference to the facet `CartesianAxisTicksFacet`.

**Comment by dcleao** [comment]
👍

**Comment by nantunes** [comment]
@dcleao: `AnyDiscreteCartesianAxis` has `DiscreteCartesianAxis` as base. Does it inherit the facets or it must be added to both?

**Comment by nantunes** [comment]
Changes done. Please complete review.

**Pull request merged**